### PR TITLE
Introduce JUnit5

### PR DESCRIPTION
**Pull request by RoiEXLab**
See #935 and #951 for more information.

This PR introduces the JUnit5 Runner now that eclipse officially supports JUnit5 in the latest 4.7.1a build.
This means you are highly encouraged to write new tests using the new namespace (importing the test annotation from junit5), and migrate the junit 4 test cases every now and then.

This is still experimental, it is working for me, but I really want to hear some feedback...

### CODE REVIEW

**Comment by RoiEXLab** [comment]
It seems that the current setup does not integrate with the jacocotestreport really well.
Do not merge this PR until we find a solution for this.

**Comment by ssoloff** [comment]
> It seems that the current setup does not integrate with the jacocotestreport really well.

Is junit-team/junit5#1024 related?

File: build.gradle [hunk 1]
```diff
@@ -11,6 +20,7 @@ plugins {
 }
 
 apply from: 'gradle/scripts/yaml.gradle'
+apply plugin: 'org.junit.platform.gradle.plugin'
```
Comment (ssoloff) [inline build.gradle]: Ugh, but at least the JUnit team is aware of the issue: junit-team/junit5#768.  However, it appears to have been pushed back to 5.1 as of 14 hours ago. 😞 

Note that there's a [workaround](https://github.com/junit-team/junit5/issues/768#issuecomment-330078905) listed in the issue, but I'm not sure if it's worth introducing that just to keep `build.gradle` clean.

Nit: Pasting the `buildscript` configuration from the JUnit docs introduced mixed tabs/spaces into this file.

**Comment by DanVanAtta** [comment]
It's really good to see your tenacity to get this working @RoiEXLab. It's valuable to get to a newer technology when we can.

**Comment by RoiEXLab** [comment]
> Is junit-team/junit5#1024 related?

@ssoloff
I think so, seems like it.

**Comment by RoiEXLab** [comment]
@ssoloff 
junit-team/junit5#1024 provides a workaround for this issue, but I have no idea how to apply it, pasting it into the buildscript isn't working 😅 
```
afterEvaluate {
    val junitPlatformTest : JavaExec by tasks
    jacoco {
        applyTo(junitPlatformTest)
    }
    task<JacocoReport>("jacocoJunit5TestReport") {
        executionData(junitPlatformTest)
        sourceSets(java.sourceSets["main"])
        sourceDirectories = files(java.sourceSets["main"].allSource.srcDirs)
        classDirectories = files(java.sourceSets["main"].output)
    }
}
```
I think this is kotlin code, any suggestions on how to apply it?
You should have push access to this branch, so in case a fix is really obvious there's no need to create a PR

**Comment by ssoloff** [comment]
@RoiEXLab I pushed a commit to your branch that implements the workaround in the issue you referenced.

Some other things I noted while playing around here:

1. As @DanVanAtta said, tests are running MUCH slower.
1. We need to suppress/redirect stdout/stderr to ensure we don't blow up the Travis logs.  Ideally, we want the same behavior as before where stdout/stderr are captured in the test results stored under `build/test-results/junit-platform/TEST-*.xml`.
1. We _should_ generate a JUnit HTML report in addition to the XML report, as was done with JUnit 4.  This makes it easier for devs to review detailed results when running tests from the command line.
1. I _think_ all the `Enclosed` tests are being run twice.  Probably because the hack we added to ensure they only run once doesn't apply to the new `junitPlatformTest` task.  I believe JUnit 5 has built-in support for nested tests, so it may be better to just convert them all incrementally.

I suspect (2) and (3) probably just require changing the `junitPlatformTest` task configuration, as needed.  And a quick fix for (4), until we can convert the tests, may just require a similar configuration change:

```
junitPlatformTest {
    exclude '**/*Tests.class'
}
```

But, unfortunately, I agree with @DanVanAtta that (1) is kind of a show-stopper until we can figure out how to improve the performance.  Maybe it's just due to the fact that the vast majority of the tests have to run through the vintage test runner, so there's extra indirection there?  Dunno...

**Comment by RoiEXLab** [comment]
@ssoloff It's possible to update all tests at once to compile with junit 5 with some basic find and replace. (replacing imports is sufficient)
I didn't do that because junit5 has reversed parameters for assertEquals and my regex knowledge is not advanced enough to automatically change that as well.
We could get a more accurate benchmark this way.

**Comment by ssoloff** [comment]
> I didn't do that because junit5 has reversed parameters for assertEquals

According to the [JUnit 5 Javadocs](http://junit.org/junit5/docs/current/api/org/junit/jupiter/api/Assertions.html#assertEquals-java.lang.Object-java.lang.Object-), the order is _expected_ followed by _actual_, which is the same as in [JUnit 4](http://junit.sourceforge.net/javadoc/org/junit/Assert.html#assertEquals(java.lang.Object,%20java.lang.Object)), and I believe has been that way  since at least JUnit 3.  Are you talking about something different?

_We_ may be using the API incorrectly (I've observed that in a few places), but it will be the same incorrect usage between JUnit 4 and 5.  The only real problem is that if an assertion fails, the message printed will contain reversed values for _expected_ and _actual_, and we can just fix those incrementally as we revisit tests.

**Comment by ssoloff** [comment]
@RoiEXLab

For issue (4), I pushed another commit that prevents the `Enclosed` tests from running twice.

For issue (2), there appears to be no way to redirect stdout/stderr at this time (see junit-team/junit5#780). 😞 

For issue (3), there is a [known limitation](https://github.com/junit-team/junit5-samples/tree/master/junit5-gradle-consumer#current-limitations) that prevents the JUnit 5 runner's results from being included in the standard Gradle test report. 😞

**Comment by RoiEXLab** [comment]
> According to the [JUnit 5 Javadocs](http://junit.org/junit5/docs/current/api/org/junit/jupiter/api/Assertions.html#assertEquals-java.lang.Object-java.lang.Object-), the order is _expected_ followed by _actual_, which is the same as in [JUnit 4](http://junit.sourceforge.net/javadoc/org/junit/Assert.html#assertEquals(java.lang.Object,%20java.lang.Object)), and I believe has been that way  since at least JUnit 3.  Are you talking about something different?

@ssoloff Interesing I could swear they were reversed at some point.

In this case I will upgrade them at once.

**Comment by DanVanAtta** [comment]
Isn't there a hamcrest style API/library built into Junit5 that avoids the expected and actual ordering problem by being type safe? It is one of the perhaps really nice features of hamcrest that you do not have to worry about it so much.

**Comment by DanVanAtta** [comment]
Running all tests (no integration) from IntelliJ, got the following results (all in seconds):
```
19.653
18.301
17.944
17.964
18.286
```

I switched over to Junit5 and did the same but only three times with the following results:
```
3m 0.880s
3m 34.805s
2m 56s
```

I remember when we had Junit5 previously for a bit that it was a bit slower, but not so drastically. Previously as well IntelliJ was running Junit5 by essentially running the build script and reporting back the results. Checking gradle builds as well on my system, the tests seem to still be running slow. I'm afraid this might be a nail in the coffin for the moment.

**Comment by RoiEXLab** [comment]
There's no longer an assertThat method by JUnit 5.
We have to use the one by hamcrest itself.
I believe (and hope) this performance issue is caused by the junit 4/5 duality and will disappear once we migrated everything, I will check...

**Comment by RoiEXLab** [comment]
I updated all tests to JUnit5.

@DanVanAtta I ran the tests, and on my machine they run equally fast using JUnit 5 and 4 (maximum 3 seconds difference), would be nice if you could verify the same.

Now there are just 2 (maybe 3) issues left.
For one the integration tests are not getting picked up. The integTest task runs fine, but does nothing.
And it seems about 150 tests less are being executed, I'd assume they are not getting picked up due to file filters or similar, some investigation will be needed.

For whatever reason javadoc errors for the integTests showed up, I have no idea why, shouldn't the test classes just get ignored?
@ssoloff Ideas?

**Comment by DanVanAtta** [comment]
good work, can confirm performance is there again:

```
gradle via CLI first run
1st run: 1m 20s total, but: "Test run finished after 26777 ms:""
2nd run: Test run finished after 24203 ms


intellij runs:
18s 614 
18s 278
18s 149
```

**Comment by ssoloff** [comment]
> For one the integration tests are not getting picked up. The integTest task runs fine, but does nothing.

I'm guessing this is due to the JUnit 5 test task no longer being of type `Test`.  We'll have to look into if there is a new type provided by the JUnit 5 Gradle plugin that we should extend for `integTest`.  Worst case, we can always keep the integration tests as JUnit 4 tests for now.

**Comment by ssoloff** [comment]
> And it seems about 150 tests less are being executed, I'd assume they are not getting picked up due to file filters or similar, some investigation will be needed.

My experience seems to be different.  Running on `master` (which this branch is currently up-to-date with minus one commit that has no code changes), I get a unit test count of 924.  Running on this branch, I get a unit test count of 926.  Not sure where those extra two tests are coming from.  I've been unsuccessful in locating them so far.  I'm going to have to just parse the test results looking for the differences (which is made more difficult by the fact that JUnit 4 puts each test class's results in a separate file, while JUnit 5 puts them all in a single file).

File: src/test/java/games/strategy/engine/config/lobby/LobbyPropertyReaderTest.java [hunk 2]
```diff
@@ -7,24 +7,24 @@
...
 
   /**
    * Sets up an example lobby property file with some fake values, then sets up a property reader test object
    * pointed at this file.
    */
-  @Before
+  @BeforeEach
   public void setup() throws IOException {
-    final File testFile = tempFolderRule.newFile();
+    final File testFile = tempFolderRule.newFile("testname");
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/config/lobby/LobbyPropertyReaderTest.java]: Recommend changing the argument to `getClass().getName()`.

File: src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java [hunk 3]
```diff
@@ -66,9 +67,9 @@
     /**
      * Sets up the test fixture.
      */
-    @Before
+    @BeforeEach
     public void setUp() throws Exception {
-      file = temporaryFolder.newFile();
+      file = temporaryFolder.newFile("filename1123");
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/framework/map/download/DownloadUtilsTests.java]: Recommend changing the argument to `getClass().getName()`.

File: src/test/java/games/strategy/engine/framework/map/download/FileSystemStrategyTest.java [hunk 4]
```diff
@@ -20,19 +21,20 @@
...
 
   private File mapFile;
 
-  @Before
+  @BeforeEach
   public void setUp() throws Exception {
     testObj = new FileSystemAccessStrategy();
     final String text = DownloadFileProperties.VERSION_PROPERTY + " = 1.2";
-    mapFile = temporaryFolder.newFile();
+    mapFile = temporaryFolder.newFile("someothertestname");
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/framework/map/download/FileSystemStrategyTest.java]: Recommend changing the argument to `getClass().getName()`.

File: src/test/java/games/strategy/test/EqualsPredicateTests.java [hunk 5]
```diff
@@ -105,7 +108,8 @@ public void shouldReturnFalseWhenObjectsAreNotEqual() {
           new FakeNestedClass(new FakeClass(-42))));
     }
 
-    private static final class FakeNestedClass {
+    @Nested
+    private final class FakeNestedClass {
```
Comment (ssoloff) [inline src/test/java/games/strategy/test/EqualsPredicateTests.java]: This class does not contain tests, and thus does not need to be annotated with `@Nested`.

File: src/test/java/games/strategy/test/EqualsPredicateTests.java [hunk 6]
```diff
@@ -159,7 +164,8 @@ public void shouldReturnFalseWhenOwnerObjectsAreNotEqual() {
       assertFalse(equalsPredicate.test(owner1, owner2));
     }
 
-    private static final class FakeOwnerClass {
+    @Nested
+    private final class FakeOwnerClass {
```
Comment (ssoloff) [inline src/test/java/games/strategy/test/EqualsPredicateTests.java]: This class does not contain tests, and thus does not need to be annotated with `@Nested`.

File: src/test/java/games/strategy/test/EqualsPredicateTests.java [hunk 7]
```diff
@@ -168,7 +174,8 @@ public void shouldReturnFalseWhenOwnerObjectsAreNotEqual() {
       }
     }
 
-    private static final class FakeOwneeClass {
+    @Nested
+    private final class FakeOwneeClass {
```
Comment (ssoloff) [inline src/test/java/games/strategy/test/EqualsPredicateTests.java]: This class does not contain tests, and thus does not need to be annotated with `@Nested`.

File: src/test/java/games/strategy/test/EqualsPredicateTests.java [hunk 8]
```diff
@@ -179,7 +186,8 @@ public void shouldReturnFalseWhenOwnerObjectsAreNotEqual() {
     }
   }
 
-  private static final class FakeClass {
+  @Nested
+  private final class FakeClass {
```
Comment (ssoloff) [inline src/test/java/games/strategy/test/EqualsPredicateTests.java]: This class does not contain tests, and thus does not need to be annotated with `@Nested`.

It also can remain `static` because it's a child of the top-level class.

**Comment by ssoloff** [comment]
> For whatever reason javadoc errors for the integTests showed up, I have no idea why, shouldn't the test classes just get ignored?

You need to update the JavadocMethod rule configuration to recognize the new JUnit 5 annotations.  In _config/checkstyle/checkstyle.xml_, you need to change this line:

```xml
<property name="allowedAnnotations" value="Override, Test, BeforeClass, AfterClass, Before, After"/>
```

**Comment by ssoloff** [comment]
> My experience seems to be different. Running on master (which this branch is currently up-to-date with minus one commit that has no code changes), I get a unit test count of 924. Running on this branch, I get a unit test count of 926.

Found the differences:

<hr>

The following top-level tests in `g.s.triplea.delegate.MatchesTests` are run in JUnit 5 but are not being run in JUnit 4:

* `testAlways`
* `testCountMatches`
* `testGetMatches`
* `testGetNMatches`
* `testGetNMatches_ShouldThrowExceptionWhenMaxIsNegative`
* `testNever`

They are not run in JUnit 4 because the top-level class is declared `Enclosed`, which means that all tests must be in a nested class.  After converting the nested classes to `Nested` in JUnit 5, these tests are now being correctly discovered and run.

<hr>

The remaining differences are due to an erroneous `@Test` annotation on the `g.s.triplea.delegate.DelegateTest#setUp()` method (introduced prior to this PR).  Because this class has four subclasses, JUnit 4 reports five occurrences of this "test."  For whatever reason, JUnit 5 only reports one occurrence of this "test."  This results in a net increase of four tests for the JUnit 4 run.

@RoiEXLab Please change `@Test` on this `setUp()` method to `@BeforeEach` as part of this PR:

```diff
diff --git a/src/test/java/games/strategy/triplea/delegate/DelegateTest.java b/src/test/java/games/strategy/triplea/delegate/DelegateTest.java
index e4d12c4..1a5378d 100644
--- a/src/test/java/games/strategy/triplea/delegate/DelegateTest.java
+++ b/src/test/java/games/strategy/triplea/delegate/DelegateTest.java
@@ -3,6 +3,7 @@ package games.strategy.triplea.delegate;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertNull;
 
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
 import games.strategy.engine.data.GameData;
@@ -78,7 +79,7 @@ public class DelegateTest {
   protected UnitType carrier;
   protected Resource pus;
 
-  @Test
+  @BeforeEach
   public void setUp() throws Exception {
     gameData = TestMapGameData.DELEGATE_TEST.getGameData();
     british = GameDataTestUtil.british(gameData);
```

<hr>

Thus +6 tests for JUnit 5 and +4 tests for JUnit 4 ends up being +2 tests for JUnit 5, which is 926 - 924.

**Comment by RoiEXLab** [comment]
@ssoloff I looked into the integTest problem.
I noticed pretty fast that there's currently no way to define an additional Task to be a junit5 task.
However there's a way to execute it in a seperate task using the junit5 console launcher api.
It's essentially an API with a main method, which automatically runs test in the specified directories.
I couldn't get it to really work, but I was able to invoke the console launcher so far:
```
diff --git a/build.gradle b/build.gradle
index cd9a86a..bad0c12 100644
--- a/build.gradle
+++ b/build.gradle
@@ -59,7 +59,7 @@ sourceSets {
         java.srcDir 'src/integ_test/java'
         resources.srcDir 'src/integ_test/resources'

-        compileClasspath = sourceSets.main.output + sourceSets.test.output + configurations.testRuntime
+        compileClasspath = sourceSets.main.output + sourceSets.test.output + configurations.testRuntime + configurations.junitPlatform
         runtimeClasspath = output + compileClasspath
     }
 }
@@ -111,19 +111,22 @@ dependencies {
     testRuntime 'org.junit.jupiter:junit-jupiter-engine:5.0.1'
 }

-task integTest(type: Test) {
+task integTest(type: junitPlatformTest.getClass()) {
     group = LifecycleBasePlugin.VERIFICATION_GROUP
     description = 'Runs the integration tests.'

-    testClassesDirs = sourceSets.integTest.output.classesDirs
     classpath = sourceSets.integTest.runtimeClasspath

-    binResultsDir = file("$buildDir/integration_test_results/binary/integTest")
+    //binResultsDir = file("$buildDir/integration_test_results/binary/integTest")

-    reports {
+    /*reports {
         reports.html.destination = file("${reports.html.destination}/$name")
         reports.junitXml.destination = file("${reports.junitXml.destination}/$name")
-    }
+    }*/
+    outputs.dir file("$buildDir/integration_test_results/binary/integTest")
+    main = org.junit.platform.console.ConsoleLauncher.class.getName()
+
+    args = ['-d', sourceSets.integTest.java.srcDirs]

     mustRunAfter tasks.test
 }
```
TL;DR There's no easy way to do it, but there is a way. See https://github.com/junit-team/junit5/issues/874 for more information

File: src/integ_test/java/games/strategy/engine/chat/ChatIntegrationTest.java [hunk 9]
```diff
@@ -79,28 +81,30 @@ public void tearDown() {
...
-
-    client1.shutdown();
-    client2.shutdown();
-    waitFor(this::clientNodesToDisconnect);
-
-    controller.deactivate();
-    waitFor(this::serverNodeToDisconnect);
+  @Test
+  public void shouldBeAbleToChatAcrossMultipleNodes() {
+    assertTimeout(Duration.ofSeconds(15), () -> {
```
Comment (RoiEXLab) [inline src/integ_test/java/games/strategy/engine/chat/ChatIntegrationTest.java]: There's a assertTimeoutPreemptively method, which fails instantly after the timeout limit is reached.
This method always lets the test methods finish first. Should we use the other method?

Comment (RoiEXLab) [reply src/integ_test/java/games/strategy/engine/chat/ChatIntegrationTest.java]: I just learned that this is the Junit4 behaviour... Changing that now
https://github.com/junit-team/junit4/wiki/Timeout-for-tests

**Comment by RoiEXLab** [comment]
When using JUnit 5, we can drop the catch exception library as JUnit 5 has a nice exception API on it's own.
```java
NullPointerException e = assertThrows(NullPointerException.java, () -> Preconditions.checkNotNull(null));
assertEquals("something", e.getMessage());
//or something like that
```

File: src/integ_test/java/games/strategy/engine/chat/ChatIntegrationTest.java (hunk 9 above)
Comment (ssoloff) [reply src/integ_test/java/games/strategy/engine/chat/ChatIntegrationTest.java]: Agreed,  It looks like the rule-of-thumb is to use `assertTimeout()` for things like performance tests (e.g. this code that never blocks needs to run within a certain amount of time), while `assertTimeoutPreemptively()` should be used for anything that can possibly block.

**Comment by ssoloff** [comment]
> When using JUnit 5, we can drop the catch exception library as JUnit 5 has a nice exception API on it's own.

Yup, I believe you mentioned that a few months ago, so I've been looking forward to switching to JUnit 5. :smile:  This PR is getting pretty big, though.  It might be easier to just do that in a follow-up PR, if you'd like.

File: build.gradle [hunk 10]
```diff
@@ -95,15 +105,12 @@ dependencies {
     testCompile 'eu.codearte.catch-exception:catch-exception:2.0.0-ALPHA-1'
     testCompile 'nl.jqno.equalsverifier:equalsverifier:2.3.3'
     testCompile 'org.hamcrest:java-hamcrest:2.0.0.0'
-    testCompile 'org.mockito:mockito-core:2.10.0'
-    testCompile 'junit:junit:4.12'
+    testCompile 'org.mockito:mockito-core:2.8.47'
```
Comment (ssoloff) [inline build.gradle]: BTW, was just curious what triggered the need to downgrade Mockito?

File: src/test/java/games/strategy/triplea/delegate/PlaceDelegateTest.java [hunk 11]
```diff
@@ -28,10 +28,8 @@
     return gameData.getUnitTypeList().getUnitType(Constants.UNIT_TYPE_INFANTRY).create(count, player);
   }
 
-  @Override
-  @Before
-  public void setUp() throws Exception {
-    super.setUp();
+  @BeforeEach
+  public void setupPlaceDelegate() {
```
Comment (ssoloff) [inline src/test/java/games/strategy/triplea/delegate/PlaceDelegateTest.java]: Does JUnit 5 make guarantees on the order `@BeforeEach` methods will be called?  That is, does it ensure that a `@BeforeEach` method in the superclass will be called before a `@BeforeEach` method in a subclass?

**EDIT:** Nevermind, answered my own question:

> `@BeforeEach` methods are inherited from superclasses as long as they are not overridden. Furthermore, `@BeforeEach` methods from superclasses will be executed before `@BeforeEach` methods in subclasses.

Comment (RoiEXLab) [reply src/test/java/games/strategy/triplea/delegate/PlaceDelegateTest.java]: Yes, I should have provided a link to the JUnit docs

File: build.gradle (hunk 10 above)
Comment (RoiEXLab) [reply build.gradle]: Bad merge conflict resolution, will fix that ^^

**Comment by RoiEXLab** [comment]
> Yup, I believe you mentioned that a few months ago, so I've been looking forward to switching to JUnit 5. :smile:  This PR is getting pretty big, though.  It might be easier to just do that in a follow-up PR, if you'd like.

@ssoloff Of course, I planned to do that in a follow-up PR, just wanted to mention it.

**Comment by RoiEXLab** [comment]
@ssoloff @DanVanAtta I figured out how to run our integrationTests with junit 5.
Worth noting is that we now explicitly require integration tests to match `^.*IntegrationTests?$` (otherwhise the integrationTest task runs the normal tests as well, because they are included in the classpath as well)

This PR is now ready for a final review, I'd prefer minor issues to be adressed in a follo-up PR though, to avoid merge conflicts as the diff is pretty large

**Comment by ssoloff** [comment]
@RoiEXLab I missed your request to defer minor issues until a follow-up PR.  I consider my suggestions to be minor.  If you want to wait, that's fine with me.  I'll change my review to Approved, but leave the PR open for @DanVanAtta.

**Comment by RoiEXLab** [comment]
This PR has been merged.
For eclipse users this means they have to update to Oxygen.1A (4.7.1a), to have JUnit 5 support from within the IDE (otherwhise they have to use gradle to do it: `gradlew clean junitPlatformTest`)

**Comment by ssoloff** [comment]
> otherwhise they have to use gradle to do it: `gradlew clean junitPlatformTest`

Note that you can still run `gradlew clean test`.  The JUnit 5 Gradle plugin apparently configures the `junitPlatformTest` task to be a dependency of the `test` task.
